src/App.tsx

```diff
@@ -15,36 +16,87 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  Default = 'default',
+  Alphabetical = 'alphabetical',
+  Length = 'length',
+}
+
 export const App: React.FC = () => {
+  const [goods, setGoods] = useState<string[]>([...goodsFromServer]);
+  const [sortOrder, setSortOrder] = useState<SortType>(SortType.Default);
+  const [isReversed, setIsReversed] = useState<boolean>(false);
+
+  const applySort = (sortType: SortType) => {
+    const sortedGoods = [...goodsFromServer];
+
+    if (sortType === SortType.Alphabetical) {
+      sortedGoods.sort((a, b) => a.localeCompare(b));
+    } else if (sortType === SortType.Length) {
+      sortedGoods.sort((a, b) => a.length - b.length);
+    }
+
+    setGoods(isReversed ? sortedGoods.reverse() : sortedGoods);
+    setSortOrder(sortType);
+  };
+
+  const handleReverse = () => {
+    setGoods(prevGoods => [...prevGoods].reverse());
+    setIsReversed(!isReversed);
+  };
+
+  const handleReset = () => {
+    setGoods([...goodsFromServer]);
+    setSortOrder(SortType.Default);
+    setIsReversed(false);
+  };
+
+  const isResetButtonVisible = sortOrder !== SortType.Default || isReversed;
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={cn('button is-info', {
+            'is-light': sortOrder !== SortType.Alphabetical,
+          })}
+          onClick={() => applySort(SortType.Alphabetical)}
+        >
           Sort alphabetically
         </button>
-
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={cn('button is-success', {
+            'is-light': sortOrder !== SortType.Length,
+          })}
+          onClick={() => applySort(SortType.Length)}
+        >
           Sort by length
         </button>
-
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={cn('button is-warning', { 'is-light': !isReversed })}
+          onClick={handleReverse}
+        >
           Reverse
         </button>
-
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {isResetButtonVisible && (
+          <button
+            type="button"
+            className="button is-danger"
+            onClick={handleReset}
+          >
+            Reset
+          </button>
+        )}
       </div>
-
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {goods.map(good => (
+          <li key={good} data-cy="Good">
+            {good}
+          </li>
```

Comment:
Using the item itself as the key in a list can lead to issues if there are duplicate items in the array. Consider using a unique identifier for each item, such as an index, if the items are not guaranteed to be unique.